docs: :memo: include comparison of languages in Why Python post (#50)
seedcase-project/decisions: why-python.qmd

title: "Why Python"
description: |
  Python is one of the most common and widely used programming languages.
  As it is used across multiple domains and industries, it is likely to be familiar to a wider audience.

> date: "2024-05-10"

Author: I was supposed to update this, right?

Reviewer: Yes.

> date: "2024-05-10"

Reviewer 2: ```suggestion
date: "2023-03-22"
date-modified: "2024-05-10"
```

::: {.column}
#### Benefits

- Well-suited for real-time and performance-critical applications.
- Allows fine-grained control over memory and hardware resources, enabling low-level, use-case-specific optimisations.

> - Compilation to native machine code for the target platform prior to execution guarantees better runtime performance than Python and, potentially and by a smaller margin, Java.

Reviewer 2: ```suggestion
- Compilation to native machine code for the target platform prior to execution guarantees better runtime performance than Python and, potentially by a smaller margin, Java.
```

::: {.column}
#### Benefits

- Code is executed by a Python interpreter, making Python programs platform independent (provided the host has an interpreter installed).
- Concise, easy-to-read, beginner-friendly syntax.
- Easy to iterate and prototype rapidly.

> - Large ecosystem of mature libraries and frameworks for web development (e.g. [Django](https://www.djangoproject.com), [Flask](https://flask.palletsprojects.com/en/2.3.x/)), database management (e.g. [SQLAlchemy](https://www.sqlalchemy.org), Django's ORM), and data analysis.

Reviewer 2: ```suggestion
- Large ecosystem of mature libraries and frameworks for web development (e.g. [Django](https://www.djangoproject.com), [Flask](https://flask.palletsprojects.com/en/2.3.x/)), database management (e.g. [SQLAlchemy](https://www.sqlalchemy.org), Django's Object Relational Mapper), and data analysis.
```

::: {.column}
#### Benefits

- Designed for statistical computing with excellent data visualisation capabilities.

> - Wide array of libraries for data analysis.

Reviewer 2: ```suggestion
- Wide array of packages for data analysis.
```

REASONS."
:::

We have decided to use Python as the main development language for Seedcase because it is a powerful, flexible and easy-to-use platform for building a data management web application. Drawing on the skills of the core Seedcase team, Python offers the best balance of capabilities and ease of development out of the options considered. It is, moreover, a language likely to be familiar to technologically-oriented Seedcase users and prospective contributors.

> While Java and C++ offer more scope for context-specific performance optimisation, we are unlikely to need this level of control for Seedcase, because the number of concurrent user requests is expected to be low, resource-intensive backend tasks are expected to run without much competition, and we don't expect to do heavy-duty, real-time data analysis or image processing. Therefore, for our use case, the advantages of these languages don't justify their steeper learning curve and added complexity.

Author: Here I'm thinking that Seedcase won't serve thousands of user requests at once, won't run very many high-workload scheduled jobs in the background and won't do super complex calculations on data. But that's just my understanding so far, so let me know if that's not the case!

Reviewer: Honestly sounds about right, but maybe @lwjohnst86 would like to comment as well.

Reviewer 2: @martonvago you got it!!
